src/rebar_git_resource.erl

```diff
@@ -107,35 +107,72 @@ download(Dir, {git, Url, ""}, State) ->
+    git_clone(rev, git_vsn(), Url, Dir, Rev).
+
+%% Use different git clone commands depending on git --version
+git_clone(branch,Vsn,Url,Dir,Branch) when Vsn >= {1,8,0} ->
+    rebar_utils:sh(?FMT("git clone ~s ~s -b ~s --single-branch",
+                        [rebar_utils:escape_chars(Url),
+                         rebar_utils:escape_chars(filename:basename(Dir)),
+                         rebar_utils:escape_chars(Branch)]),
+                   [{cd, filename:dirname(Dir)}]);
+git_clone(branch,_Vsn,Url,Dir,Branch) ->
+    rebar_utils:sh(?FMT("git clone ~s ~s -b ~s",
+                        [rebar_utils:escape_chars(Url),
+                         rebar_utils:escape_chars(filename:basename(Dir)),
+                         rebar_utils:escape_chars(Branch)]),
+                   [{cd, filename:dirname(Dir)}]);
+git_clone(tag,Vsn,Url,Dir,Tag) when Vsn >= {1,8,0} ->
+    rebar_utils:sh(?FMT("git clone ~s ~s -b ~s --single-branch",
+                        [rebar_utils:escape_chars(Url),
+                         rebar_utils:escape_chars(filename:basename(Dir)),
+                         rebar_utils:escape_chars(Tag)]),
+                   [{cd, filename:dirname(Dir)}]);
+git_clone(tag,_Vsn,Url,Dir,Tag) ->
+    rebar_utils:sh(?FMT("git clone ~s ~s -b ~s",
+                        [rebar_utils:escape_chars(Url),
+                         rebar_utils:escape_chars(filename:basename(Dir)),
+                         rebar_utils:escape_chars(Tag)]),
+                   [{cd, filename:dirname(Dir)}]);
+git_clone(ref,_Vsn,Url,Dir,Ref) ->
     rebar_utils:sh(?FMT("git clone -n ~s ~s",
                         [rebar_utils:escape_chars(Url),
                          rebar_utils:escape_chars(filename:basename(Dir))]),
                    [{cd, filename:dirname(Dir)}]),
     rebar_utils:sh(?FMT("git checkout -q ~s", [Ref]), [{cd, Dir}]);
-download(Dir, {git, Url, Rev}, _State) ->
-    ?WARN("WARNING: It is recommended to use {branch, Name}, {tag, Tag} or {ref, Ref}, otherwise updating the dep may not work as expected.", []),
-    ok = filelib:ensure_dir(Dir),
+git_clone(rev,_Vsn,Url,Dir,Rev) ->
     rebar_utils:sh(?FMT("git clone -n ~s ~s",
                         [rebar_utils:escape_chars(Url),
                          rebar_utils:escape_chars(filename:basename(Dir))]),
                    [{cd, filename:dirname(Dir)}]),
     rebar_utils:sh(?FMT("git checkout -q ~s", [rebar_utils:escape_chars(Rev)]),
                    [{cd, Dir}]).
 
+git_vsn() ->
+    case rebar_utils:sh("git --version",[]) of
+        {ok, VsnStr} ->
+            case re:run(VsnStr, "git version\\h+(\\d)\\.(\\d)\\.(\\d).*",[{capture,[1,2,3],list}]) of
+                {match,[Maj,Min,Patch]} ->
+                    {list_to_integer(Maj),
+                     list_to_integer(Min),
+                     list_to_integer(Patch)};
+                nomatch ->
+                    undefined
+            end;
+        {error,_} ->
+            undefined
+    end.
```

Comment:
Should we be able to cache the result? I'm not too sure about using the pdict or the app env; on some OSes forking out is always expensive, so avoiding it may not be bad.